File: DataStructuresAndAlgorithms/1. Array and Hashing/4. Group Anagram/groupAnagrams.cpp
#include <bits/stdc++.h>
#include <vector>
#include <unordered_map>

using namespace std;

vector<vector<string>> groupAnagrams(vector<string> strs){
    vector<string> sortStrs;
    copy(strs.begin(), strs.end(), back_inserter(sortStrs));
    for(int i = 0; i < sortStrs.size(); i++){
        sort(sortStrs[i].begin(), sortStrs[i].end());
    }
    unordered_map<string, vector<string>> groups;
    for(int i = 0; i < strs.size(); i++){
        groups[sortStrs[i]].push_back(strs[i]);
    }
    vector<vector<string>> ans;
    int j = 0;
    for(auto x: groups){
        vector<string> ans2;
        for(int i = 0; i < x.second.size(); i++){
            ans2.push_back(x.second[i]);
        }
        ans.push_back(ans2);
        j++;
    }   
    return ans;
}


int main(int argc, char **argv) {
    
    vector<string> strs = {"eat","tea","tan","ate","nat","bat"};
    vector<vector<string>> groups;
    groups = groupAnagrams(strs);
    for(int i=0; i < groups.size(); i++) {
        for(int j=0; j < groups[i].size(); j++) {
            cout<< groups[i][j] << " ";
        }
        cout<<endl;
    }

    return 0;
}

File: DataStructuresAndAlgorithms/1. Array and Hashing/4. Group Anagram/groupAnagrams.py
def groupAnagrams(strs: list) -> list:
        sort_strs = ["".join(sorted(s)) for s in strs]
        
        hash_table = {}
        
        for i in range(len(strs)):
            
            if sort_strs[i] in hash_table:
                hash_table[sort_strs[i]].append(strs[i])
            else:
                hash_table[sort_strs[i]] = [strs[i]]
        
        return hash_table.values()


if __name__ == "__main__":
    strs = ["eat","tea","tan","ate","nat","bat"]
    # [""]
    # ["a"]
    print(groupAnagrams(strs))